src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7577,126 +7452,109 @@ private ArrayRankSpecifierSyntax ParseArrayRankSpecifier(out bool sawNonOmittedS
+                var element = ParseTupleElement();
+                list.Add(element);
 
-                    while (this.CurrentToken.Kind == SyntaxKind.CommaToken)
-                    {
-                        var comma = this.EatToken(SyntaxKind.CommaToken);
-                        list.AddSeparator(comma);
+                while (this.CurrentToken.Kind == SyntaxKind.CommaToken)
+                {
+                    var comma = this.EatToken(SyntaxKind.CommaToken);
+                    list.AddSeparator(comma);
 
-                        element = ParseTupleElement();
-                        list.Add(element);
-                    }
+                    element = ParseTupleElement();
+                    list.Add(element);
                 }
+            }
 
-                if (list.Count < 2)
+            if (list.Count < 2)
+            {
+                if (list.Count < 1)
                 {
-                    if (list.Count < 1)
-                    {
-                        list.Add(_syntaxFactory.TupleElement(this.CreateMissingIdentifierName(), identifier: null));
-                    }
-
-                    list.AddSeparator(SyntaxFactory.MissingToken(SyntaxKind.CommaToken));
-                    var missing = this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_TupleTooFewElements);
-                    list.Add(_syntaxFactory.TupleElement(missing, identifier: null));
+                    list.Add(_syntaxFactory.TupleElement(this.CreateMissingIdentifierName(), identifier: null));
                 }
 
-                var close = this.EatToken(SyntaxKind.CloseParenToken);
-                var result = _syntaxFactory.TupleType(open, list, close);
+                list.AddSeparator(SyntaxFactory.MissingToken(SyntaxKind.CommaToken));
+                var missing = this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_TupleTooFewElements);
+                list.Add(_syntaxFactory.TupleElement(missing, identifier: null));
+            }
 
-                result = CheckFeatureAvailability(result, MessageID.IDS_FeatureTuples);
+            var result = _syntaxFactory.TupleType(
+                open,
+                _pool.ToListAndFree(list),
+                this.EatToken(SyntaxKind.CloseParenToken));
 
-                return result;
-            }
-            finally
-            {
-                _pool.Free(list);
-            }
+            return CheckFeatureAvailability(result, MessageID.IDS_FeatureTuples);
         }
 
         private TupleElementSyntax ParseTupleElement()
         {
-            var type = ParseType();
-            SyntaxToken name = null;
-
-            if (IsTrueIdentifier())
-            {
-                name = this.ParseIdentifierToken();
-            }
-
-            return _syntaxFactory.TupleElement(type, name);
+            return _syntaxFactory.TupleElement(
+                ParseType(),
+                IsTrueIdentifier() ? this.ParseIdentifierToken() : null);
```

Comment:
use of ternaries allows us to model constructs like `X -> A B? C` giving the same nice construction/recursive-descent parity/clarity.  We use this in many other places, so this is just making things consistent here.